client/asset/interface.go

```diff
@@ -379,6 +379,12 @@ type Wallet interface {
 	// EstimateRegistrationTxFee returns an estimate for the tx fee needed to
 	// pay the registration fee using the provided feeRate.
 	EstimateRegistrationTxFee(feeRate uint64) uint64
+	// EstimateSendTxFee returns a tx fee estimate for sending or withdrawing
+	// the provided amount using the provided feeRate. This uses actual utxos to
+	// calculate the tx fee where possible.
+	EstimateSendTxFee(address string, value, feeRate uint64, subtract bool) (fee uint64, isValidAddress bool, err error)
```

Comment:
This can be an optional interface.  Mainly because I'm going to do the Electrum impl. later, but I don't think this is core functionality.  The caller in `Core` can do something else sensible if the wallet doesn't support this operation.

Make optional in `asset.Wallet` interface:

```diff
diff --git a/client/asset/interface.go b/client/asset/interface.go
index ca2d8205..c76f4bff 100644
--- a/client/asset/interface.go
+++ b/client/asset/interface.go
@@ -379,12 +379,15 @@ type Wallet interface {
 	// EstimateRegistrationTxFee returns an estimate for the tx fee needed to
 	// pay the registration fee using the provided feeRate.
 	EstimateRegistrationTxFee(feeRate uint64) uint64
+	// ValidateAddress checks that the provided address is valid.
+	ValidateAddress(address string) bool
+}
+
+type TxFeeEstimator interface {
 	// EstimateSendTxFee returns a tx fee estimate for sending or withdrawing
 	// the provided amount using the provided feeRate. This uses actual utxos to
 	// calculate the tx fee where possible.
 	EstimateSendTxFee(address string, value, feeRate uint64, subtract bool) (fee uint64, isValidAddress bool, err error)
-	// ValidateAddress checks that the provided address is valid.
-	ValidateAddress(address string) bool
 }
 
 // Rescanner is a wallet implementation with rescan functionality.
```

Move from `btc.Wallet` top optional interface:

```diff
diff --git a/client/asset/btc/wallet.go b/client/asset/btc/wallet.go
index b8edf202..51bb4d09 100644
--- a/client/asset/btc/wallet.go
+++ b/client/asset/btc/wallet.go
@@ -49,7 +49,6 @@ type Wallet interface {
 	ownsAddress(addr btcutil.Address) (bool, error) // this should probably just take a string
 	getWalletTransaction(txHash *chainhash.Hash) (*GetTransactionResult, error)
 	reconfigure(walletCfg *asset.WalletConfig, currentAddress string) (restartRequired bool, err error)
-	estimateSendTxFee(tx *wire.MsgTx, feeRate uint64, subtract bool) (fee uint64, err error)
 }
 
 type tipRedemptionWallet interface {
@@ -61,6 +60,10 @@ type tipRedemptionWallet interface {
 	findRedemptionsInMempool(ctx context.Context, reqs map[outPoint]*findRedemptionReq) (discovered map[outPoint]*findRedemptionResult)
 }
 
+type txFeeEstimator interface {
+	estimateSendTxFee(tx *wire.MsgTx, feeRate uint64, subtract bool) (fee uint64, err error)
+}
+
 // walletTxChecker provide a fast wallet tx query when block info not needed.
 // This should be treated as an optimization method, where getWalletTransaction
 // may always be used in its place if it does not exists.
```

Require the above optional interface in `intermediaryWallet` only, not `baseWallet` (which is what `ExchangeWalletElectrum` embeds`):

```diff
diff --git a/client/asset/btc/btc.go b/client/asset/btc/btc.go
index 6e56b919..17bce33a 100644
--- a/client/asset/btc/btc.go
+++ b/client/asset/btc/btc.go
@@ -754,6 +754,7 @@ func (w *baseWallet) useSplitTx() bool {
 
 type intermediaryWallet struct {
 	*baseWallet
+	txFeeEstimator
 	tipRedeemer tipRedemptionWallet
 }
 
@@ -4365,29 +4366,31 @@ func (btc *baseWallet) createWitnessSig(tx *wire.MsgTx, idx int, pkScript []byte
 
 // EstimateSendTxFee returns a tx fee estimate for sending or withdrawing the
 // provided amount using the provided feeRate.
-func (btc *baseWallet) EstimateSendTxFee(address string, sendAmount, feeRate uint64, subtract bool) (fee uint64, isValidAddress bool, err error) {
+func (btc *intermediaryWallet) EstimateSendTxFee(address string, sendAmount, feeRate uint64, subtract bool) (fee uint64, isValidAddress bool, err error) {
 	if sendAmount == 0 {
 		return 0, false, fmt.Errorf("cannot check fee: send amount = 0")
 	}
 
@@ -4397,7 +4400,7 @@ func (btc *baseWallet) EstimateSendTxFee(address string, sendAmount, feeRate uin
 
 	tx := wire.NewMsgTx(wire.TxVersion)
 	tx.AddTxOut(wireOP)
-	fee, err = btc.node.estimateSendTxFee(tx, btc.feeRateWithFallback(feeRate), subtract)
+	fee, err = btc.estimateSendTxFee(tx, btc.feeRateWithFallback(feeRate), subtract)
 	if err != nil {
 		return 0, isValidAddress, err
 	}
```

Reverting the Electrum changes:

```diff
diff --git a/client/asset/btc/electrum/wallet_methods.go b/client/asset/btc/electrum/wallet_methods.go
index f2a7fc35..e6feb85c 100644
--- a/client/asset/btc/electrum/wallet_methods.go
+++ b/client/asset/btc/electrum/wallet_methods.go
@@ -10,8 +10,6 @@ import (
 	"fmt"
 	"strconv"
 	"strings"
-
-	"github.com/btcsuite/btcd/btcjson"
 )
 
 const (
@@ -280,15 +278,6 @@ func (wc *WalletClient) GetBalance(ctx context.Context) (*Balance, error) {
 	}, nil
 }
 
-func (wc *WalletClient) EstimateSendTxFee(ctx context.Context, txHex string, options interface{}, segwit bool) (fee float64, err error) {
-	res := &btcjson.FundRawTransactionResult{}
-	err = wc.Call(ctx, methodFundRawTransaction, positional{txHex, options, segwit}, &res)
-	if err != nil {
-		return 0, fmt.Errorf("error calculating transaction fee: %w", err)
-	}
-	return res.Fee.ToBTC(), nil
-}
-
 // payto(self, destination, amount, fee=None, feerate=None, from_addr=None, from_coins=None, change_addr=None,
 // nocheck=False, unsigned=False, rbf=None, password=None, locktime=None, addtransaction=False, wallet: Abstract_Wallet = None):
 type paytoReq struct {
diff --git a/client/asset/btc/electrum_client.go b/client/asset/btc/electrum_client.go
index d7dbcc5e..11f648f7 100644
--- a/client/asset/btc/electrum_client.go
+++ b/client/asset/btc/electrum_client.go
@@ -60,7 +60,7 @@ type electrumWalletClient interface {
 	GetRawTransaction(ctx context.Context, txid string) ([]byte, error) // wallet method
 	GetAddressHistory(ctx context.Context, addr string) ([]*electrum.GetAddressHistoryResult, error)
 	GetAddressUnspent(ctx context.Context, addr string) ([]*electrum.GetAddressUnspentResult, error)
-	EstimateSendTxFee(ctx context.Context, txHex string, options interface{}, segwit bool) (fee float64, err error)
 }
 
 type electrumNetworkClient interface {
@@ -1255,36 +1255,6 @@ func (ew *electrumWallet) sendToAddress(address string, value, feeRate uint64, s
 	return chainhash.NewHashFromStr(txid)
 }
 
-// part of btc.Wallet interface.
-func (ew *electrumWallet) estimateSendTxFee(tx *wire.MsgTx, feeRate uint64, subtract bool) (fee uint64, err error) {
-	txBytes, err := ew.serializeTx(tx)
-	if err != nil {
-		return 0, fmt.Errorf("tx serialization error: %w", err)
-	}
-
-	// 1e-5 = 1e-8 for satoshis * 1000 for kB.
-	feeRateOption := float64(feeRate) / 1e5
-	options := &btcjson.FundRawTransactionOpts{
-		FeeRate:    &feeRateOption,
-		ChangeType: &btcjson.ChangeTypeBech32,
-	}
-	if subtract {
-		options.SubtractFeeFromOutputs = []int{0}
-	}
-	if !ew.segwit {
-		options.ChangeType = &btcjson.ChangeTypeLegacy
-	}
-
-	txFee, err := ew.wallet.EstimateSendTxFee(ew.ctx, hex.EncodeToString(txBytes), options, false)
-	if err != nil {
-		return 0, err
-	}
-
-	fee = toSatoshi(txFee)
-
-	return
-}
-
 func (ew *electrumWallet) sweep(ctx context.Context, address string, feeRate uint64) ([]byte, error) {
 	txRaw, err := ew.wallet.Sweep(ctx, ew.walletPass(), address, float64(feeRate))
 	if err != nil {
```

Reply:
Thanks for the review.